pkg/operator/controller/route-metrics/controller.go

```diff
@@ -0,0 +1,180 @@
+	// Cast the received object into Route object.
+	routeObject := obj.(*routev1.Route)
+
+	// Iterate through the related Ingress Controllers.
+	for _, ic := range routeObject.Status.Ingress {
+		log.Info("queueing ingresscontroller", "name", ic.RouterName)
+		// Create a reconcile.Request for the Ingress Controller.
+		request := reconcile.Request{
+			NamespacedName: types.NamespacedName{
+				Name:      ic.RouterName,
+				Namespace: r.Namespace,
+			},
+		}
+		requests = append(requests, request)
+	}
+	return requests
+}
+
+// reconciler handles the actual ingresscontroller reconciliation logic in response to events.
+type reconciler struct {
+	client    client.Client
+	Namespace string
+}
+
+// Reconcile expects request to refer to a Ingress Controller resource, and will do all the work to gather metrics related to
+// the resource.
+func (r *reconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	log.Info("reconciling", "request", request)
+
+	// Fetch the Ingress Controller object.
+	ingressController := &operatorv1.IngressController{}
+	if err := r.client.Get(ctx, request.NamespacedName, ingressController); err != nil {
+		if kerrors.IsNotFound(err) {
+			// This means the Ingress Controller object was already deleted/finalized.
+			log.Info("Ingress Controller not found", "request", request)
+			// Delete the Shard label corresponding to the Ingress Controller from the RouteMetricsControllerRoutesPerShard metric.
+			DeleteRouteMetricsControllerRoutesPerShardMetric(request.Name)
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, fmt.Errorf("failed to get Ingress Controller %q: %v", request, err)
+	}
+
+	// List all the Namespaces which matches the Namespace LabelSelector.
+	namespaceList := corev1.NamespaceList{}
+	if ingressController.Spec.NamespaceSelector != nil {
+		namespaceLabelSelector := ingressController.Spec.NamespaceSelector.MatchLabels
+		if err := r.client.List(ctx, &namespaceList, &client.ListOptions{
+			LabelSelector: labels.SelectorFromSet(namespaceLabelSelector),
+		}); err != nil {
+			return reconcile.Result{}, fmt.Errorf("failed to list Namespaces %q: %v", request, err)
+		}
+	} else {
+		// If Namespace LabelSelector is not provided then add Namespace with empty name to indicate all Namespaces.
+		namespaceList.Items = append(namespaceList.Items, corev1.Namespace{})
+	}
+
+	// Get the Route LabelSelector from the Ingress Controller.
+	var routeLabelSelector map[string]string
+	if ingressController.Spec.RouteSelector != nil {
+		routeLabelSelector = ingressController.Spec.RouteSelector.MatchLabels
+	}
+
+	// Store the previous value of the metric.
+	prevMetricValue := GetRouteMetricsControllerRoutesPerShardMetric(request.Name)
+
+	// Initialize the RouteMetricsControllerRoutesPerShard metric for Ingress Controller.
+	InitializeRouteMetricsControllerRoutesPerShardMetric(request.Name)
+
+	// Iterate through all the Namespaces.
```

Comment:
Does this work if there is no Namespace LabelSelector?  How about if there is no routeLabelSelector?  Or both are missing?

Reply:
This case is handled by adding an empty namespace object to the `NamespaceList.Items` slice (line 126).

If  routeLabelSelector is missing then it would not have any effect, as it will fetch the Routes without using any selector.

Reply:
Please add unit tests for the three scenarios.